File: _OLD/packages/portfolio-api/src/routes/secrets.ts
import { Elysia, t } from "elysia";
import { resetPriceService } from "../services/price-service-factory";

const SERVICE = "com.pebble.desktop";

async function keytarSet(name: string, value: string): Promise<void> {
  const { default: keytar } = await import("keytar");
  await keytar.setPassword(SERVICE, name, value);
}

async function keytarDelete(name: string): Promise<void> {
  const { default: keytar } = await import("keytar");
  await keytar.deletePassword(SERVICE, name);
}

export const secretPlugin = new Elysia({ prefix: "/api/secrets" })
  .post("/:name", async ({ params, body }) => {
    const name = params.name;
    const { value } = body;
    try {
      await keytarSet(name, value);
    } catch {
      // Fallback: keytar not available in plain dev mode
    }
    process.env[`SECRET_${name.toUpperCase().replace(/-/g, "_")}`] = value;
    if (name === "coingecko-api-key") {
      process.env["COINGECKO_API_KEY"] = value;
      resetPriceService();
    }
    return { ok: true };
  }, {
    body: t.Object({
      value: t.String(),
    })
  })
  .delete("/:name", async ({ params }) => {
    const name = params.name;
    try { await keytarDelete(name); } catch { /* ignore */ }
    delete process.env[`SECRET_${name.toUpperCase().replace(/-/g, "_")}`];
    if (name === "coingecko-api-key") {
      delete process.env["COINGECKO_API_KEY"];
      resetPriceService();
    }
    return { ok: true };
  });

File: fastapi/app/api/routes/secrets.py
import os

from fastapi import APIRouter

from app.models import SecretSet
from app.services.price_service_factory import reset_price_service

router = APIRouter(prefix="/secrets", tags=["secrets"])

_KEYRING_SERVICE = "com.portfolio.desktop"


def _keyring_set(name: str, value: str) -> None:
    try:
        import keyring
        keyring.set_password(_KEYRING_SERVICE, name, value)
    except Exception:
        pass


def _keyring_delete(name: str) -> None:
    try:
        import keyring
        keyring.delete_password(_KEYRING_SERVICE, name)
    except Exception:
        pass


@router.post("/{name}")
def set_secret(name: str, body: SecretSet) -> dict:
    _keyring_set(name, body.value)
    env_key = f"SECRET_{name.upper().replace('-', '_')}"
    os.environ[env_key] = body.value
    if name == "coingecko-api-key":
        os.environ["COINGECKO_API_KEY"] = body.value
        reset_price_service()
    return {"ok": True}


@router.delete("/{name}")
def delete_secret(name: str) -> dict:
    _keyring_delete(name)
    env_key = f"SECRET_{name.upper().replace('-', '_')}"
    os.environ.pop(env_key, None)
    if name == "coingecko-api-key":
        os.environ.pop("COINGECKO_API_KEY", None)
        reset_price_service()
    return {"ok": True}
